### PR TITLE
Trim quotes for non-string values in nested json parsing

### LINKED ISSUE #11817: [BUG] Allow string to numeric conversion with "dtype" argument in JSON reader
**Describe the bug**
The `cudf.read_json` does not correctly convert from strings to numeric type when `engine='cudf_experimental'` is set and the `dtype` argument specifies a numeric type. `engine='cudf'` and pandas both support strings to numeric casting.

**Steps/Code to reproduce bug**
```
>>> cudf.read_json(StringIO('{"k": "1"}'), engine='cudf_experimental', lines=True, dtype={'k': 'int'})
   k
0  0
>>> cudf.read_json(StringIO('{"k": "1"}'), engine='cudf', lines=True, dtype={'k': 'int'})
   k
0  1
>>> pd.read_json(StringIO('{"k": "1"}'), lines=True, dtype={'k': 'int'})
   k
0  1
```

**Expected behavior**
I expect the string data to be converted into the specified numeric dtype when possible, and to be null when it cannot be cast.

**Environment overview (please complete the following information)**
docker image "rapidsai/rapidsai-dev-nightly", "22.10-cuda11.5-devel-ubuntu18.04-py3.8", "418d71a6792c"
cudf built from source on commit `9e9ba6e8518cf36c42fdbdd60276bd414444827e`

**Additional context**
The behavior when some of the strings cannot be cast is currently not consistent. For now this issue addresses the case where all the strings cast as valid integers, but we may want to broaden the scope.
```
>>> pd.read_json(StringIO('{"k": "1"}\n{"k":"a"}'), lines=True, dtype={'k': 'int'})
   k
0  1
1  a
>>> cudf.read_json(StringIO('{"k": "1"}\n{"k":"a"}'), lines=True, dtype={'k': 'int'})
    k
0   1
1  10
>>> cudf.read_json(StringIO('{"k": "1"}\n{"k":"a"}'), lines=True, engine='cudf_experimental', dtype={'k': 'int'})
   k
0  0
1  0
```

### PR DESCRIPTION
**Pull request by karthikeyann**
## Description
Trim quotes for non-string values in nested json parsing
Added corner cases for unquoted and quoted literals. (Review the unit test)
Fixes old json reader to treat `"null"` as string instead of NULL.

closes #11817

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by karthikeyann** [comment]
rerun tests

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (vyasr) [inline cpp/src/io/utilities/parsing_utils.cuh:514]:
```suggestion
 * @param[in] end Pointer to the first character after the parsing range
```

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (vyasr) [inline cpp/src/io/utilities/parsing_utils.cuh:555]: Correct me if I'm wrong, it seems like in dropping `decode_value` we lost support for duration and timestamp types. Is that intentional, is it OK because those weren't actually supported before and those code paths were never taken, or is it an intentional breaking change?

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (vyasr) [inline cpp/src/io/utilities/parsing_utils.cuh:577]: Is this TODO intended to be a question of what we should consider invalid?

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (vyasr) [reply cpp/src/io/utilities/parsing_utils.cuh:555]: Ah it looks like the branches where we could hit duration or timestamp types have had the logic inlined below, so maybe this is fine then.

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (vyasr) [inline cpp/src/io/utilities/parsing_utils.cuh:653]: It looks like this is where the timestamp/duration overloads for `decode_value` would have hit before, so now it's just inlining that logic. Is that right?

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (vyasr) [inline cpp/src/io/utilities/parsing_utils.cuh:632]: I'm confused. The PR description says that we switched from returning NaN to returning null, but here it looks like we're still inserting a `quiet_NaN`. What am I missing?

**Review by vyasr — COMMENTED** [review]
I'm not familiar with this code, so I have a few questions that may have very obvious answers. Overall the changes look OK though.

File: cpp/src/io/json/json_gpu.cu
Comment (bdice) [inline cpp/src/io/json/json_gpu.cu:259]: Is it always safe to dereference `*(desc.value_begin - 1)`? I didn't fully grasp `next_field_descriptor` on a quick glance, but I don't see why the value's start couldn't be at the beginning of the input.

File: cpp/src/io/json/json_gpu.cu
Comment (bdice) [inline cpp/src/io/json/json_gpu.cu:267]: Don't add/subtract/multiply a raw boolean. `static_cast` to a pointer difference type.

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (bdice) [inline cpp/src/io/utilities/parsing_utils.cuh:515]: Only use `@param[in]` if there is a `@param[out]`. Otherwise just use `@param`.

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (bdice) [reply cpp/src/io/utilities/parsing_utils.cuh:577]: If this involves changes to fixed-point handling and isn't strictly related to trimming quotes, let's revert the changes here and file an issue or defer to another PR.

File: cpp/tests/io/json_test.cpp
Comment (bdice) [inline cpp/tests/io/json_test.cpp:1537]: Is this a leftover debug print?

**Review by bdice — CHANGES_REQUESTED** [review]
I'm having a bit of difficulty with this PR's scope -- there are a lot of changes that do not appear related to the title (trimming quotes in nested JSON parsing). The change log line generated from this PR's title will not explain all the unrelated bug fixes for CSV parsing, the old JSON reader, cleanup/refactoring, and changes in data type parsing. There are some bits of this PR that are breaking, and some bits that are nonbreaking. Can we split this work into a coherent set of PRs that have separated, well-labeled changes?

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (karthikeyann) [reply cpp/src/io/utilities/parsing_utils.cuh:653]: Yes. It seemed like another level of indirection only for timestamp, duration types. Rest of the types had SFINAE anyway. So, used `if constexpr` directly here.

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (karthikeyann) [reply cpp/src/io/utilities/parsing_utils.cuh:632]: This `quiet_NaN` write need not happen if this row is null.  When the row is null, the value held in respective memory location does not matter. But it's writing the value anyway. I felt the reason to leave `NaN` value in the memory location because float has a `NaN` representation, but other types doesn't. I will remove this.

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (karthikeyann) [reply cpp/src/io/utilities/parsing_utils.cuh:577]: I will split the changes to multiple PRs. There are 3 fixes in this PR.

File: cpp/src/io/json/json_gpu.cu
Comment (karthikeyann) [reply cpp/src/io/json/json_gpu.cu:259]: `next_field_descriptor` strips off quotes and whitespaces.
I had the same doubt as well. I kept this because I saw a similar line at 
https://github.com/karthikeyann/cudf/blob/f38240316bcea8a40e0c160c47a82b1d64556479/cpp/src/io/json/json_gpu.cu#L348

You are right. It isn't safe to deference it, so, I will fix both issues in this PR. (Already this PR got multiple minor bug fixes!)

File: cpp/tests/io/json_test.cpp
Comment (karthikeyann) [reply cpp/tests/io/json_test.cpp]: removed.

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (karthikeyann) [reply cpp/src/io/utilities/parsing_utils.cuh]: updated.

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (karthikeyann) [reply cpp/src/io/utilities/parsing_utils.cuh]: moved to https://github.com/rapidsai/cudf/pull/12022

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (karthikeyann) [reply cpp/src/io/utilities/parsing_utils.cuh]: moved this code to PR https://github.com/rapidsai/cudf/pull/12022 and updated as per this comment

**Comment by karthikeyann** [comment]
Moved unrelated bug fixes to separate PRs #12021 and #12022

File: cpp/src/io/json/json_gpu.cu
Comment (karthikeyann) [reply cpp/src/io/json/json_gpu.cu:279]: fixed.

File: cpp/src/io/json/json_gpu.cu
Comment (karthikeyann) [reply cpp/src/io/json/json_gpu.cu]: fixed.

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (karthikeyann) [reply cpp/src/io/utilities/parsing_utils.cuh]: moved to  #12022

File: cpp/src/io/json/json_gpu.cu
Comment (karthikeyann) [reply cpp/src/io/json/json_gpu.cu]: FYI: It is safe for results of `next_field_descriptor` because it will atleast have `:` as the previous character.

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (karthikeyann) [reply cpp/src/io/utilities/parsing_utils.cuh]: @vyasr CSV_TEST
`CsvReaderTest.InvalidFloatingPoint` expects the value to be nan regardless of bitmask.
https://github.com/karthikeyann/cudf/blob/fc5b67517765f5ebc94abead4b4b6c8cbd43d4da/cpp/tests/io/csv_test.cpp#L1191-L1194

Should this test be updated to ignore the values, and check only bitmask?

Comment (vyasr) [reply cpp/src/io/utilities/parsing_utils.cuh]: Yes I think so. Checking the data in a null row is UB.

File: cpp/src/io/json/json_gpu.cu
Comment (vyasr) [inline cpp/src/io/json/json_gpu.cu:205]: To make @bdice happy :) : `thrust::distance` will return an `int`, not a bool, and you are also doing arithmetic with this after, so maybe just save it as an `int` named `dist_to_first_non_quote` or something explicit.

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (vyasr) [inline cpp/src/io/utilities/parsing_utils.cuh:557]: Am I crazy, or is this function not used anywhere now? It looks like you have currently inlined the logic for determining `is_quoted` (and it's a one-line thrust call, so I'm fine with that), making this function obsolete.

**Review by vyasr — APPROVED** [review]
Thanks for splitting this PR up, it is much easier to review now. I've got a couple of cleanup requests, but they are easy enough to handle that I don't think this requires another review from me. Feel free to re-request if you end up needing to make more substantial changes for any reason.

**Comment by vyasr** [comment]
@karthikeyann is this PR still considered breaking, or is this just a bugfix now (with the breaking changes in one of the other PRs)?

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (karthikeyann) [reply cpp/src/io/utilities/parsing_utils.cuh:557]: It's used in `cpp/include/cudf/io/detail/data_casting.cuh`.

File: cpp/src/io/json/json_gpu.cu
Comment (karthikeyann) [reply cpp/src/io/json/json_gpu.cu:205]: `thrust::distance(..) >=2` is the expression, which returns bool. `thrust::distance` checks the size of the string.
For arithmetic with point, I can static_cast to `ptrdiff_t`.

Comment (karthikeyann) [reply cpp/src/io/json/json_gpu.cu:205]: updated.

**Comment by karthikeyann** [comment]
> @karthikeyann is this PR still considered breaking, or is this just a bugfix now (with the breaking changes in one of the other PRs)?

Old json reader behaviour change is still a breaking change.

**Comment by karthikeyann** [comment]
rerun tests

File: cpp/src/io/json/json_gpu.cu
Comment (vyasr) [reply cpp/src/io/json/json_gpu.cu:205]: Thanks, changes look right.

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (bdice) [inline cpp/src/io/utilities/parsing_utils.cuh:530]: If there are no output parameters, only mark `@param` instead of `@param[in]` and `@param[out]`.
```suggestion
 * @param begin Pointer to the first character in the parsing range
 * @param end Pointer to the first character after the parsing range
```

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (bdice) [inline cpp/src/io/utilities/parsing_utils.cuh:559]: We've been discussing avoiding default parameters in detail APIs (discussed on Slack, with initial changes in PRs like #12007 but no policy change in the docs yet). Let's drop the default here, especially since it is not used. The calling site provides an explicit value.
```suggestion
                                                                      char quotechar)
```

File: cpp/src/io/utilities/parsing_utils.cuh
Comment (bdice) [inline cpp/src/io/utilities/parsing_utils.cuh:553]: See comment below about avoiding detail parameter defaults.
```suggestion
 * @param quotechar The character used to denote quotes. Provide '\0' if no quotes should be trimmed.
```

**Review by bdice — APPROVED** [review]
Minor suggestions. Thank you very much for splitting this work into multiple PRs. This was far easier to understand as a complete unit.

**Review by vuule — COMMENTED** [review]
Looks good, bit concerned about the test.

File: python/cudf/cudf/tests/test_json.py
Comment (karthikeyann) [reply python/cudf/cudf/tests/test_json.py:739]: Great suggestion. Compared it. 
pandas has different behaviour than both cudf engine.

I think pandas infers types and parses the json, and then converts to the user specified type. Because of this, there are some difference in output.
1) All int literals:
For example, for json string `[{"k": 1}, {"k": 2}, {"k": null}, {"k": 4}]` 
`pd.read_json(StringIO(jsonstr), orient="records", dtype={"k": "str"})` creates output
```
     k
0  1.0
1  2.0
2  NaN
3  4.0
```
while both cudf engines create output
```
      k
0     1
1     2
2  <NA>
3     4
```

2) int and `"` string literals mixed
If quotes are present, even if there are literals and user specified type is `int`, it parses as string. This seems  very odd behaviour compared to previous example.

Comment (karthikeyann) [reply python/cudf/cudf/tests/test_json.py:739]: Should we deviate from pandas behaviour, or replicate the pandas behaviour in cudf?

Comment (vuule) [reply python/cudf/cudf/tests/test_json.py:739]: Oof. To put it politely, our behavior makes more sense to me. Let's keep it at least until a customer asks for different behavior :D 
I still wish the test would compare with ground truth. Maybe add a test with one or few cases where the output is compared with hard-coded values?

Comment (karthikeyann) [reply python/cudf/cudf/tests/test_json.py:739]: added ground truth test cases.

**Comment by karthikeyann** [comment]
@gpucibot merge

**Pull request merged**
